### PR TITLE
Sl feature update

### LINKED ISSUE #188: Include all symmetries during training
https://github.com/MaMiFreak/RocAlphaGo/pull/5/files

i made an implementation that includes all symmetries in the shuffle file
instead of noop/diag2 i used 0/8 or the shuffle file will become to big

my questions
- do we want the option to remove certain symmetries? or would true/false suffice to include all or none
- i added a max-validation argument that is able to override the --train-val-test fraction setting, agree?
- test set is not used, do we want to do anything with it?
- any concerns on the code used or improvements?

### PR DESCRIPTION
**Pull request by MaMiFreak**
Fix #188 #149 #148 
Keras 1.2.0 does not accept initial batch for some reason, unsure why it is not included
all TODO's refer to improve explaining texts please read them and give feedback when they do not give a good explanation.

Theano backend checks required:
- test changing lr 
- correct weight names

### CODE REVIEW

File: AlphaGo/training/supervised_policy_trainer.py [hunk 1]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
...
+
+
 class MetadataWriterCallback(Callback):
+    """Set current batch at training start
+       Save metadata and Model after every epoch
+    """
 
-    def __init__(self, path):
+    def __init__(self, path, root, metadata):
+        super(Callback, self).__init__()
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: `super(MetaDataWriterCallback, self).__init__()`

File: AlphaGo/training/supervised_policy_trainer.py [hunk 2]
```diff
@@ -158,110 +260,520 @@ def run_training(cmd_line_args=None):
...
-        args.minibatch,
-        symmetries)
+        metadata["batch_size"])
 
-    sgd = SGD(lr=args.learning_rate, decay=args.decay)
-    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=["accuracy"])
+    # check if step decay has to be applied
+    if metadata["decay_every"] is None:
+        # use normal decay without momentum
+        lr_scheduler_callback = LrDecayCallback(metadata["learning_rate"],
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: Since `LrDecayCallback` appears to work the same as `SGD`'s built-in decay, this could be simplified. Here you could do `sgd = SGD(lr=metadata["learning_rate"], decay=metadata["decay"])`, and in the `else` you could create `SGD(...,decay=0)` and set up the `LRStepDecayCallback`.

File: AlphaGo/training/supervised_policy_trainer.py [hunk 3]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
...
+        if ans not in ['y', 'Y', 'n', 'N']:
+            print 'please enter y or n.'
+            continue
+        if ans == 'y' or ans == 'Y':
+            return True
+        if ans == 'n' or ans == 'N':
+            return False
+
+
+class LrDecayCallback(Callback):
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: We can simplify by using the built-in decay of `SGD` instead of creating a new class here.

File: AlphaGo/training/supervised_policy_trainer.py [hunk 4]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
...
+        self.model.optimizer.current_batch += 1
+
+
+class LrStepDecayCallback(Callback):
+    """Set learning rate every decay_every batches according to:
+       initial_learning_rate * (decay ^ (current_batch / decay every))
+    """
+
+    def __init__(self, learning_rate, decay_every, decay, verbose):
+        super(Callback, self).__init__()
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: `super(LrStepDecayCallback, self).__init__()`

File: AlphaGo/training/supervised_policy_trainer.py [hunk 5]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
...
+        self.verbose = verbose
+        self.decay = decay
+
+    def set_lr(self):
+        # calculate learning rate
+        n_decay = int(self.model.optimizer.current_batch / self.decay_every)
+        new_lr = self.learning_rate * (self.decay ** n_decay)
+
+        # set new learning rate
+        self.model.optimizer.lr = K.variable(value=new_lr)
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: Rather than create a new variable every epoch, could this be `K.set_value(self.model.optimizer.lr, new_lr)`?

File: AlphaGo/training/supervised_policy_trainer.py [hunk 6]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
...
+        # print new learning rate if verbose
+        if self.verbose:
+            print("\nBatch: " + str(self.model.optimizer.current_batch) +
+                  " New learning rate: " + str(new_lr))
+
+    def on_train_begin(self, logs={}):
+        # set initial learning rate
+        self.set_lr()
+
+    def on_batch_begin(self, batch, logs={}):
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: 1. It would simplify things - and match the AG paper - if we applied the decay `on_epoch_begin` rather than `on_batch_begin`. Batches are more general, of course, but at the cost of added complexity to this file (if it were epochs, then we wouldn't need to keep track of `current_batch`)
2. `set_lr` could be renamed to `on_XYZ_begin`

File: AlphaGo/training/supervised_policy_trainer.py [hunk 7]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
...
-    "rot90": lambda feature: np.rot90(feature, 1),
-    "rot180": lambda feature: np.rot90(feature, 2),
-    "rot270": lambda feature: np.rot90(feature, 3),
-    "fliplr": lambda feature: np.fliplr(feature),
-    "flipud": lambda feature: np.flipud(feature),
-    "diag1": lambda feature: np.transpose(feature),
-    "diag2": lambda feature: np.fliplr(np.rot90(feature, 1))
-}
+        # save model to file with correct epoch
+        save_file = os.path.join(self.root, FOLDER_WEIGHT,
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: anticipating that keras will soon fix the `initial_epoch` bug, I think we should keep the `ModelCheckpoint` separate from the `MetadataWriter`.

Proposal:
```python
class ModelCheckpointWithOffset(ModelCheckpoint):
    def __init__(self, offset, *args, **kwargs):
        super(ModelCheckpoint, self).__init__(*args, **kwargs)
        self.offset = offset

    def on_epoch_end(self, epoch, logs={}):
        super(ModelCheckpoint, self).on_epoch_end(epoch + self.offset, logs)
```

if/when keras fixes the bug, we just swap the `ModelCheckpoint` back in.

File: AlphaGo/training/supervised_policy_trainer.py [hunk 8]
```diff
@@ -2,40 +2,78 @@
...
 def one_hot_action(action, size=19):
     """Convert an (x,y) action into a size x size array of zeros with a 1 at x,y
     """
+
     categorical = np.zeros((size, size))
     categorical[action] = 1
     return categorical
 
 
 def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
```
Comment (wrongu) [inline AlphaGo/training/supervised_policy_trainer.py]: now that you're keeping track of the last batch that the model saw, should the data generator pick up where it left off on resume?

Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: i  have been checking how the data generator works because i was concerned about shuffeling data every batch and multi threading unfortunately i have not figured it out jet, although my intention is to add this if possible

File: AlphaGo/training/supervised_policy_trainer.py [hunk 9]
```diff
@@ -45,99 +83,163 @@ def shuffled_hdf5_batch_generator(state_dataset, action_dataset,
                 yield (Xbatch, Ybatch)
 
 
+def confirm(prompt=None, resp=False):
```
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: we can move but i would move it to the model script as value net uses a different batch generator

File: AlphaGo/training/supervised_policy_trainer.py (hunk 6 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: AG does apply this on batch 80.000.000 that's why i implemented it this way also if you implement it in `on_epocht_begin` you have to calculate your batch, and epoch size correctly in order to have the decay on the correct batch (difficult to explain)

File: AlphaGo/training/supervised_policy_trainer.py (hunk 3 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: i implemented this class in order to get the correct lr when restarting because initial_epoch cannot be set.

File: AlphaGo/training/supervised_policy_trainer.py (hunk 7 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: possible but the ``ModelCheckpoint` adds another callback even though it is only two lines of code in order to save weights file. 
i was thinking on changing the name to `EpochDataSaver` but forgot to do so in the end.

File: AlphaGo/training/supervised_policy_trainer.py (hunk 2 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: added comment about why i implemented separate SGD decay class

File: AlphaGo/training/supervised_policy_trainer.py (hunk 9 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: sorry, i was thinking about something else. this would definitely be something for `util.py`

File: AlphaGo/training/supervised_policy_trainer.py (hunk 5 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: yes that works, added it

File: AlphaGo/training/supervised_policy_trainer.py (hunk 2 above)
Comment (MaMiFreak) [inline AlphaGo/training/supervised_policy_trainer.py]: i check my code history and the first implementation might be interesting(at that moment it looked like it was not working but it was because of a different error, still i have to check the lr over time in order to be 100% sure)
`
# oldcode
    # check if step decay has to be applied
    if metadata["decay_every"] is None:
        # use normal decay without momentum
        # initial learning rate
        if resume:
            initial_lr = metadata["learning_rate"] * (1. / (1. + metadata["decay"] * metadata["current_batch"]))
        else:
            initial_lr = metadata["learning_rate"]
        sgd = SGD(lr=initial_lr, decay=metadata["decay"])`

but i am not sure this would be easier to understand
